src/coins.cpp

```diff
@@ -1557,6 +1576,65 @@ CValidationState::Code CCoinsViewCache::IsScTxApplicableToState(const CTransacti
     return CValidationState::Code::OK;
 }
 
+#ifdef ENABLE_ADDRESS_INDEXING
+void CCoinsViewCache::HandleIndexesSidechainEvents(int height, CBlockTreeDB* pblocktree,
+                                                   std::vector<std::pair<CAddressIndexKey, CAddressIndexValue>>& addressIndex,
+                                                   std::vector<std::pair<CAddressUnspentKey, CAddressUnspentValue>>& addressUnspentIndex)
+{
+    if (!HaveSidechainEvents(height))
+        return;
+
+    CSidechainEvents scEvents;
+    GetSidechainEvents(height, scEvents);
+
+    //Handle Ceasing Sidechain
+    for (const uint256& ceasingScId : scEvents.ceasingScs)
+    {
+        CSidechain sidechain;
+        assert(GetSidechain(ceasingScId, sidechain));
+
+        if (sidechain.lastTopQualityCertReferencedEpoch == CScCertificate::EPOCH_NULL) {
+            assert(sidechain.lastTopQualityCertHash.IsNull());
+            continue;
+        }
+
+        CTxIndexValue txIndexVal;
+        assert(pblocktree->ReadTxIndex(sidechain.lastTopQualityCertHash, txIndexVal));
+
+        // Set the lower quality BTs as superseded
+        UpdateBackwardTransferIndexes(sidechain.lastTopQualityCertHash, txIndexVal.txIndex, addressIndex, addressUnspentIndex,
+                                      CCoinsViewCache::flagIndexesUpdateType::SUPERSEDE_CERTIFICATE);
+    }
+}
+
+void CCoinsViewCache::RevertIndexesSidechainEvents(int height, CBlockUndo& blockUndo, CBlockTreeDB* pblocktree,
+                                                   std::vector<std::pair<CAddressIndexKey, CAddressIndexValue>>& addressIndex,
+                                                   std::vector<std::pair<CAddressUnspentKey, CAddressUnspentValue>>& addressUnspentIndex)
+{
+    if (!HaveSidechainEvents(height))
+        return;
+
+    // Reverting ceasing sidechains
+    for (auto it = blockUndo.scUndoDatabyScId.begin(); it != blockUndo.scUndoDatabyScId.end(); ++it)
+    {
+        if ((it->second.contentBitMask & CSidechainUndoData::AvailableSections::CEASED_CERT_DATA) == 0)
+            continue;
+
+        const uint256& scId = it->first;
+        const CSidechain* const pSidechain = AccessSidechain(scId);
+
+        if (pSidechain->lastTopQualityCertReferencedEpoch != CScCertificate::EPOCH_NULL)
+        {
+            CTxIndexValue txIndexVal;
+            assert(pblocktree->ReadTxIndex(pSidechain->lastTopQualityCertHash, txIndexVal));
+
+            // Set the old top quality BTs as valid (even not mature yet)
+            UpdateBackwardTransferIndexes(pSidechain->lastTopQualityCertHash, txIndexVal.txIndex, addressIndex, addressUnspentIndex,
+                                          CCoinsViewCache::flagIndexesUpdateType::RESTORE_CERTIFICATE);
+        }
```

Comment:
Update also txIndex setting the lastTopQualCertHash with certMaturityHeight positive